flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -1155,58 +1156,78 @@ struct XEmboxOpConversion : public EmboxCommonConversion<fir::cg::XEmboxOp> {
-    assert(xbox.shape().size() && "must have a shape");
-    unsigned shapeOff = 1;
-    bool hasShift = xbox.shift().size();
-    unsigned shiftOff = shapeOff + xbox.shape().size();
-    bool hasSlice = xbox.slice().size();
-    unsigned sliceOff = shiftOff + xbox.shift().size();
+    auto base = operands[0];
+    assert(!xbox.shape().empty() && "must have a shape");
+    unsigned shapeOff = xbox.shapeOffset();
+    bool hasShift = !xbox.shift().empty();
+    unsigned shiftOff = xbox.shiftOffset();
+    bool hasSlice = !xbox.slice().empty();
+    unsigned sliceOff = xbox.sliceOffset();
     auto loc = xbox.getLoc();
     mlir::Value zero = genConstantIndex(loc, i64Ty, rewriter, 0);
     mlir::Value one = genConstantIndex(loc, i64Ty, rewriter, 1);
     mlir::Value prevDim = integerCast(loc, rewriter, i64Ty, eleSize);
+    mlir::Value prevPtrOff = one;
     auto eleTy = boxTy.getEleTy();
-    for (unsigned d = 0; d < rank; ++d) {
-      // store lower bound (normally 0)
-      mlir::Value lb = zero;
-      if (eleTy.isa<fir::PointerType>() || eleTy.isa<fir::HeapType>() ||
-          hasSlice) {
-        lb = one;
-        if (hasShift)
-          lb = operands[shiftOff];
-        if (hasSlice)
-          lb = rewriter.create<mlir::LLVM::SubOp>(loc, i64Ty, lb,
-                                                  operands[sliceOff]);
-      }
-      dest = insertField(rewriter, loc, dest, {7, d, 0}, lb);
-
-      // store extent
+    const auto rank = xbox.getRank();
+    for (unsigned di = 0, d = 0; di < rank; ++di) {
       mlir::Value extent = operands[shapeOff];
       mlir::Value outerExtent = extent;
-      if (hasSlice) {
-        extent = rewriter.create<mlir::LLVM::SubOp>(
-            loc, i64Ty, operands[sliceOff + 1], operands[sliceOff]);
-        extent = rewriter.create<mlir::LLVM::AddOp>(loc, i64Ty, extent,
+      if (hasSlice && mlir::isa_and_nonnull<fir::UndefOp>(
+                          xbox.slice()[3 * di + 1].getDefiningOp())) {
+        // If this dimension is invariant, add it to the base.
+        auto off = operands[sliceOff];
+        auto adj = one;
+        if (hasShift)
+          adj = operands[shiftOff];
+        auto ao = rewriter.create<mlir::LLVM::SubOp>(loc, i64Ty, off, adj);
+        auto po =
+            rewriter.create<mlir::LLVM::MulOp>(loc, i64Ty, ao, prevPtrOff);
+        base = rewriter.create<mlir::LLVM::GEPOp>(loc, base.getType(),
+                                                  mlir::ValueRange{base, po});
+      } else {
+        // store lower bound (normally 0)
+        auto lb = zero;
+        if (eleTy.isa<fir::PointerType>() || eleTy.isa<fir::HeapType>() ||
+            hasSlice) {
+          lb = one;
+          if (hasShift)
+            lb = operands[shiftOff];
+          if (hasSlice)
+            lb = rewriter.create<mlir::LLVM::SubOp>(loc, i64Ty, lb,
+                                                    operands[sliceOff]);
```

Comment:
When passing a slice, do we really care about passing the slice lower bounds as the descriptor lower bounds ? shouldn't the base address be shifted by the sum of the lb*strides instead (strides of the base array, before applying the slice extent) ?

I do not see how on the callee side we would retrieve the actual base address with the passed lower bounds, given the extent/stride in the descriptor applies to the sliced layout, and the lower bounds should be applied with the non-sliced layout.

Reply:
I think we discussed this already...